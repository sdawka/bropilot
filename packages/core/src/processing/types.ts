export interface ProcessingResult {
  sessionId?: string; // Optional for overall report
  processedMessages: number;
  extractedEntities: ExtractedEntities;
  conflicts: Conflict[];
  confidence: number;
}

export interface ExtractedEntity {
  type: string;
  name: string;
  confidence: number;
  sourceMessages: string[]; // message IDs
  attributes: Record<string, any>;
  description?: string; // Added for module generation
  suggestedFields?: {
    name: string;
    type: string;
    constraints?: any;
    required?: boolean;
  }[]; // Added for module generation
  relationships?: any[]; // Added for module generation
}

export interface ExtractedDomain extends ExtractedEntity {
  type: 'domain';
  description: string;
}

export interface JsonSchema {
  type: string;
  properties: { [key: string]: any };
  required: string[];
}

export interface ExtractedFeature extends ExtractedEntity {
  type: 'feature';
  purpose: string;
  domains: string[]; // names of domains
}

export interface ExtractedRequirement extends ExtractedEntity {
  type: 'requirement';
  description: string;
  feature?: string; // name of feature
}

export interface ExtractedEntities {
  domains: ExtractedDomain[];
  features: ExtractedFeature[];
  requirements: ExtractedRequirement[];
}

export interface Conflict {
  type: string;
  entity1: ExtractedEntity;
  entity2?: ExtractedEntity;
  resolution: 'manual' | 'auto' | 'create_domain';
  message?: string;
}

export interface ValidationResult {
  valid: boolean;
  errors: string[];
}

export interface ChatSession {
  id: string;
  name: string;
  lastProcessedMessageId: string | null;
}

export interface ChatMessage {
  id: string;
  sessionId: string;
  role: 'user' | 'assistant';
  content: string;
  timestamp: Date;
  processed: boolean;
}

export interface ProcessOptions {
  quiet: boolean;
  force: boolean;
  rollbackOnError: boolean;
}
