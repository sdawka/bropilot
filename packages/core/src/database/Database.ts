import Database from 'better-sqlite3';
import * as path from 'path';
import { fileURLToPath } from 'url';
import * as fs from 'fs';
import { v4 as uuidv4 } from 'uuid';

interface Migration {
  id: number;
  name: string;
  created_at: number;
}

export class AppDatabase {
  private db: Database.Database;
  private migrationsPath: string;

  constructor(dbPath: string = ':memory:', migrationsPath?: string) {
    this.db = new Database(dbPath);
    this.db.pragma('journal_mode = WAL');
    this.db.pragma('foreign_keys = ON');
    const __filename = fileURLToPath(import.meta.url);
    const __dirname = path.dirname(__filename);
    this.migrationsPath = migrationsPath || path.join(__dirname, 'migrations');
    this.initMigrationsTable();
  }

  private initMigrationsTable() {
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS migrations (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        name TEXT NOT NULL UNIQUE,
        created_at INTEGER NOT NULL
      );
    `);
  }

  async migrate() {
    const appliedMigrations = this.db
      .prepare('SELECT name FROM migrations')
      .all() as { name: string }[];
    const appliedMigrationNames = new Set(appliedMigrations.map((m) => m.name));

    const migrationFiles = fs
      .readdirSync(this.migrationsPath)
      .filter((file) => file.endsWith('.sql'))
      .sort();

    for (const file of migrationFiles) {
      if (!appliedMigrationNames.has(file)) {
        const migrationSql = fs.readFileSync(
          path.join(this.migrationsPath, file),
          'utf8',
        );
        try {
          this.db.exec(migrationSql);
          this.db
            .prepare('INSERT INTO migrations (name, created_at) VALUES (?, ?)')
            .run(file, Date.now());
          console.log(`Applied migration: ${file}`);
        } catch (error) {
          console.error(`Failed to apply migration ${file}:`, error);
          throw error;
        }
      }
    }
  }

  getDB(): Database.Database {
    return this.db;
  }

  close(): void {
    this.db.close();
  }

  transaction<T>(fn: (...args: any[]) => T): (...args: any[]) => T {
    const func = this.db.transaction(fn);
    return (...args: any[]) => func(...args);
  }

  // Helper for UUID generation
  generateId(): string {
    return uuidv4();
  }
}
