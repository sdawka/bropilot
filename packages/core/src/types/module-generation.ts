/**
 * Placeholder interfaces for module generation
 */
export interface ModuleGenerationOptions {
  domains?: string[];
  force?: boolean;
  interactive?: boolean;
  explain?: boolean;
}

export interface KnowledgeGraph {
  getDomainsByNames(names: string[]): Promise<Domain[]>;
  getAllDomains(): Promise<Domain[]>;
  getModuleByDomain(domainId: string): Promise<Module | null>;
  getFeaturesByDomain(domainId: string): Promise<Feature[]>;
}

export interface Domain {
  id: string;
  name: string;
  description: string;
  responsibilities: string[];
}

export interface Module {
  name: string;
  type: 'core' | 'ui';
  description: string;
  domain_id: string;
  interface: ModuleInterface;
  state: ModuleState;
  things: Thing[];
  behaviors: Behavior[];
  flows: Flow[];
  components?: Component[];
  screens?: Screen[];
}

export interface ModuleInterface {
  type: 'web_app' | 'rpc_api';
  description: string;
}

export interface ModuleState {
  type: 'nanostores' | 'postgresql';
  stores?: any[];
  schema?: string;
}

export interface Thing {
  type: 'thing';
  name: string;
  description: string;
  schema: JsonSchema;
  invariants: string[];
  relationships: any[];
}

export interface Behavior {
  type: 'behavior';
  name: string;
  description: string;
  trigger: string;
  preconditions: string[];
  rules: string[];
  effects: string[];
}

export interface Flow {
  type: 'flow';
  name: string;
  description: string;
  steps: any[];
}

export interface Component {
  type: 'component';
  name: string;
  description: string;
  componentType: string; // renamed from 'type' to avoid conflict
  props: any;
}

export interface Screen {
  name: string;
  description: string;
  components: string[];
}

export interface Feature {
  id: string;
  name: string;
  purpose: string;
  requirements: string[];
}

export interface ExtractedEntity {
  name: string;
  description: string;
  suggestedFields: {
    name: string;
    type: string;
    constraints?: any;
    required?: boolean;
  }[];
  relationships: any[];
}

export interface JsonSchema {
  type: string;
  properties: { [key: string]: any };
  required: string[];
}

// --- Additional types for code generation ---

export interface Task {
  id: string;
  name: string;
  description: string;
  // Add more fields as needed
}

export interface CodeDependency {
  name: string;
  version?: string;
  path?: string;
  // Add more fields as needed
}

export interface ProjectConfig {
  language: string;
  methodology?: string;
  ui_framework?: string;
  styling?: string;
  test_components?: boolean;
  // Add more fields as needed
}

export interface Import {
  name: string;
  path: string;
  // Add more fields as needed
}

export interface Export {
  name: string;
  type: string;
  // Add more fields as needed
}

export interface GeneratedTest {
  filePath: string;
  content: string;
  // Add more fields as needed
}
