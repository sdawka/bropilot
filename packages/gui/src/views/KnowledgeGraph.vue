<template>
  <div class="p-8">
    <h1 class="text-3xl font-bold mb-4">Knowledge Graph</h1>
    <p>Visual graph view will appear here.</p>
  </div>
</template>
<script setup></script>
