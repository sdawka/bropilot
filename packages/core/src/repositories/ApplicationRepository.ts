import Database from 'better-sqlite3';
import { BaseRepository } from './BaseRepository.js';
import { ApplicationSchema } from '../database/schema.js';

export type { ApplicationSchema as Application };

export class ApplicationRepository extends BaseRepository<ApplicationSchema> {
  constructor(db: Database.Database) {
    super(db, 'applications');
  }
}
