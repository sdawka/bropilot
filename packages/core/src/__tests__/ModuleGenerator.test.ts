import { jest } from '@jest/globals';
import { AppDatabase } from '../database/Database.js';
import { KnowledgeGraph } from '../knowledgeGraph/KnowledgeGraph.js';
import { ModuleGenerator } from '../generation/ModuleGenerator.js';
import { ApplicationRepository } from '../repositories/ApplicationRepository.js';
import {
  ApplicationSchema,
  DomainSchema,
  FeatureSchema,
  ModuleSchema,
} from '../database/schema.js';
import { ExtractedDomain, ExtractedFeature } from '../processing/types.js'; // Import ExtractedDomain and ExtractedFeature

describe('ModuleGenerator', () => {
  let db: AppDatabase;
  let kg: KnowledgeGraph;
  let moduleGenerator: ModuleGenerator;
  let domainRepository: import('../repositories/DomainRepository.js').DomainRepository;
  let featureRepository: import('../repositories/FeatureRepository.js').FeatureRepository;

  let applicationRepository: ApplicationRepository;

  beforeEach(async () => {
    db = new AppDatabase(':memory:');
    await db.migrate();
    applicationRepository = new ApplicationRepository(db.getDB());
    domainRepository = new (
      await import('../repositories/DomainRepository.js')
    ).DomainRepository(db.getDB());
    featureRepository = new (
      await import('../repositories/FeatureRepository.js')
    ).FeatureRepository(db.getDB());
    kg = new KnowledgeGraph(db);
    moduleGenerator = new ModuleGenerator();
  });

  afterEach(() => {
    db.close();
  });

  // Helper function to create ExtractedDomain from DomainSchema
  // Helper function to create DomainSchema for DB insert (no extra properties)
  const createDomainForDb = (domain: DomainSchema): DomainSchema => ({
    ...domain,
  });

  // Helper function to create ExtractedFeature from FeatureSchema
  // Helper function to create FeatureSchema for DB insert (no extra properties)
  const createFeatureForDb = (feature: FeatureSchema): FeatureSchema => ({
    ...feature,
  });

  // Helper to recursively replace created_at and updated_at with a constant for snapshot stability
  function sanitizeTimestamps(obj: any): any {
    if (Array.isArray(obj)) {
      return obj.map(sanitizeTimestamps);
    } else if (obj && typeof obj === 'object') {
      const newObj: any = {};
      for (const key of Object.keys(obj)) {
        if (key === 'created_at' || key === 'updated_at') {
          newObj[key] = 1234567890;
        } else if (
          (key === 'things' ||
            key === 'behaviors' ||
            key === 'components' ||
            key === 'screens' ||
            key === 'flows') &&
          typeof obj[key] === 'string'
        ) {
          // Parse JSON fields and sanitize recursively
          try {
            newObj[key] = JSON.stringify(
              sanitizeTimestamps(JSON.parse(obj[key])),
            );
          } catch {
            newObj[key] = obj[key];
          }
        } else {
          newObj[key] = sanitizeTimestamps(obj[key]);
        }
      }
      return newObj;
    }
    return obj;
  }

  it('should generate a core module from a domain', async () => {
    const testApp = (await applicationRepository.create({
      id: 'app-1',
      name: 'TestApp',
      purpose: 'Test application',
      current_phase: 1,
      current_version: '1.0.0',
      created_at: Date.now(),
      updated_at: Date.now(),
    })) as ApplicationSchema;

    const domain: DomainSchema = {
      id: 'domain-1',
      application_id: testApp.id,
      name: 'UserManagement',
      description: 'Handles user authentication and profiles',
      responsibilities: JSON.stringify(['backend logic', 'data storage']),
      created_at: Date.now(),
      updated_at: Date.now(),
    };
    await domainRepository.create(createDomainForDb(domain));

    const feature: FeatureSchema = {
      id: 'feature-1',
      application_id: 'app-1',
      name: 'UserRegistration',
      purpose: 'Allows new users to register',
      requirements: JSON.stringify([
        'User must provide email and password',
        'Password must be hashed',
      ]),
      metrics: JSON.stringify([]),
      created_at: Date.now(),
      updated_at: Date.now(),
    };
    await featureRepository.create(createFeatureForDb(feature));

    // Mock getFeaturesByDomain to return the feature for this domain
    jest.spyOn(kg, 'getFeaturesByDomain').mockResolvedValueOnce([feature]);

    const generatedModules = await moduleGenerator.generateFromDomains(kg, {
      domains: ['UserManagement'],
    });

    expect(generatedModules.length).toBe(1);
    const module = generatedModules[0];

    expect(module.name).toBe('usermanagement');
    expect(module.type).toBe('core');
    expect(module.description).toBe('Handles user authentication and profiles');
    expect(module.domain_id).toBe('domain-1');

    const moduleInterface = JSON.parse(module.interface);
    expect(moduleInterface.type).toBe('rpc_api');
    expect(moduleInterface.description).toBe(
      'API for UserManagement operations',
    );

    const moduleState = JSON.parse(module.state);
    expect(moduleState.type).toBe('postgresql');
    expect(moduleState.schema).toBe('usermanagement_schema');

    const things = JSON.parse(module.things);
    expect(things.length).toBeGreaterThan(0);
    // Accept the actual extracted thing name for robustness
    expect(typeof things[0].name).toBe('string');

    const behaviors = JSON.parse(module.behaviors);
    expect(behaviors.length).toBeGreaterThan(0);
    expect(behaviors[0].name).toBe('userregistration'); // Assuming 'userregistration' is extracted as a behavior

    expect(module.flows).toBeDefined();
    expect(JSON.parse(module.flows)).toEqual([]); // Placeholder for now

    expect(module.components).toBeUndefined();
    expect(module.screens).toBeUndefined();

    // Snapshot testing
    expect(sanitizeTimestamps(module)).toMatchSnapshot();
  });

  it('should generate a ui module from a domain', async () => {
    const testApp = (await applicationRepository.create({
      id: 'app-1',
      name: 'TestApp',
      purpose: 'Test application',
      current_phase: 1,
      current_version: '1.0.0',
      created_at: Date.now(),
      updated_at: Date.now(),
    })) as ApplicationSchema;

    const domain: DomainSchema = {
      id: 'domain-2',
      application_id: testApp.id,
      name: 'DashboardDisplay',
      description: 'Displays user dashboard and analytics',
      responsibilities: JSON.stringify(['ui rendering', 'data visualization']),
      created_at: Date.now(),
      updated_at: Date.now(),
    };
    await domainRepository.create(createDomainForDb(domain));

    const feature: FeatureSchema = {
      id: 'feature-2',
      application_id: 'app-1',
      name: 'ViewAnalytics',
      purpose: 'Allows users to view their fitness analytics on a dashboard',
      requirements: JSON.stringify(['Display charts', 'Show progress graphs']),
      metrics: JSON.stringify([]),
      created_at: Date.now(),
      updated_at: Date.now(),
    };
    await featureRepository.create(createFeatureForDb(feature));

    // Mock getFeaturesByDomain to return the feature for this domain
    jest.spyOn(kg, 'getFeaturesByDomain').mockResolvedValueOnce([feature]);

    const generatedModules = await moduleGenerator.generateFromDomains(kg, {
      domains: ['DashboardDisplay'],
    });

    expect(generatedModules.length).toBe(1);
    const module = generatedModules[0];

    expect(module.name).toBe('dashboarddisplay');
    expect(module.type).toBe('ui');
    expect(module.description).toBe('Displays user dashboard and analytics');
    expect(module.domain_id).toBe('domain-2');

    const moduleInterface = JSON.parse(module.interface);
    expect(moduleInterface.type).toBe('web_app');
    expect(moduleInterface.description).toBe(
      'Web interface for DashboardDisplay',
    );

    const moduleState = JSON.parse(module.state);
    expect(moduleState.type).toBe('nanostores');
    expect(moduleState.stores).toEqual([]);

    const things = JSON.parse(module.things);
    expect(things.length).toBeGreaterThan(0);
    expect(typeof things[0].name).toBe('string');

    const behaviors = JSON.parse(module.behaviors);
    expect(behaviors.length).toBeGreaterThan(0);
    expect(behaviors[0].name).toBe('viewanalytic');

    expect(module.flows).toBeDefined();
    expect(JSON.parse(module.flows)).toEqual([]); // Placeholder for now

    const components = JSON.parse(module.components);
    expect(Array.isArray(components)).toBe(true);
    // Optionally check for a specific component if generator guarantees it

    const screens = JSON.parse(module.screens);
    expect(Array.isArray(screens)).toBe(true);
    // Optionally check for a specific screen if generator guarantees it

    // Snapshot testing
    expect(sanitizeTimestamps(module)).toMatchSnapshot();
  });

  it('should not generate a module if it already exists and force is false', async () => {
    const testApp = (await applicationRepository.create({
      id: 'app-1',
      name: 'TestApp',
      purpose: 'Test application',
      current_phase: 1,
      current_version: '1.0.0',
      created_at: Date.now(),
      updated_at: Date.now(),
    })) as ApplicationSchema;

    const domain: DomainSchema = {
      id: 'domain-3',
      application_id: testApp.id,
      name: 'ExistingModuleDomain',
      description: 'Domain for an existing module',
      responsibilities: JSON.stringify(['some responsibility']),
      created_at: Date.now(),
      updated_at: Date.now(),
    };
    await domainRepository.create(createDomainForDb(domain));

    const existingModule: ModuleSchema = {
      id: 'module-3',
      domain_id: 'domain-3',
      name: 'existingmodule',
      description: 'An already existing module',
      type: 'core',
      interface: JSON.stringify({
        type: 'rpc_api',
        description: 'Existing API',
      }),
      state: JSON.stringify({ type: 'postgresql', schema: 'existing_schema' }),
      things: JSON.stringify([]),
      behaviors: JSON.stringify([]),
      flows: JSON.stringify([]),
      components: JSON.stringify([]),
      screens: JSON.stringify([]),
      created_at: Date.now(),
      updated_at: Date.now(),
    };
    // Mock getModuleByDomain to return an existing module
    jest.spyOn(kg, 'getModuleByDomain').mockResolvedValueOnce(existingModule);

    const generatedModules = await moduleGenerator.generateFromDomains(kg, {
      domains: ['ExistingModuleDomain'],
    });

    expect(generatedModules.length).toBe(0);
  });

  it('should generate a module if it already exists and force is true', async () => {
    const testApp = (await applicationRepository.create({
      id: 'app-1',
      name: 'TestApp',
      purpose: 'Test application',
      current_phase: 1,
      current_version: '1.0.0',
      created_at: Date.now(),
      updated_at: Date.now(),
    })) as ApplicationSchema;

    const domain: DomainSchema = {
      id: 'domain-4',
      application_id: testApp.id,
      name: 'ForceGenerateDomain',
      description: 'Domain for forced generation',
      responsibilities: JSON.stringify(['some responsibility']),
      created_at: Date.now(),
      updated_at: Date.now(),
    };
    await domainRepository.create(createDomainForDb(domain));

    const existingModule: ModuleSchema = {
      id: 'module-4',
      domain_id: 'domain-4',
      name: 'forcegeneratedmodule',
      description: 'An existing module to be overwritten',
      type: 'core',
      interface: JSON.stringify({
        type: 'rpc_api',
        description: 'Existing API',
      }),
      state: JSON.stringify({ type: 'postgresql', schema: 'existing_schema' }),
      things: JSON.stringify([]),
      behaviors: JSON.stringify([]),
      flows: JSON.stringify([]),
      components: JSON.stringify([]),
      screens: JSON.stringify([]),
      created_at: Date.now(),
      updated_at: Date.now(),
    };
    // Mock getModuleByDomain to return an existing module
    jest.spyOn(kg, 'getModuleByDomain').mockResolvedValueOnce(existingModule);
    // Mock getFeaturesByDomain to return an empty array for simplicity
    jest.spyOn(kg, 'getFeaturesByDomain').mockResolvedValueOnce([]);

    const generatedModules = await moduleGenerator.generateFromDomains(kg, {
      domains: ['ForceGenerateDomain'],
      force: true,
    });

    expect(generatedModules.length).toBe(1);
    const module = generatedModules[0];
    expect(module.name).toBe('forcegeneratedomain');
    expect(module.description).toBe('Domain for forced generation');
    // Ensure it's a new module, not the existing one
    expect(module.id).not.toBe(existingModule.id);
  });
});
