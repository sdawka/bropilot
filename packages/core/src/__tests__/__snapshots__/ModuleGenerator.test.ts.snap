// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleGenerator should generate a core module from a domain 1`] = `
{
  "behaviors": "[{"id":"","module_id":"","name":"userregistration","description":"Allows new users to register","trigger":"rpc_call","input_schema":"{}","output_schema":"{}","actions":"[\\"User must provide email and password\\",\\"Password must be hashed\\"]","created_at":1234567890,"updated_at":1234567890}]",
  "description": "Handles user authentication and profiles",
  "domain_id": "domain-1",
  "flows": "[]",
  "interface": "{"type":"rpc_api","description":"API for UserManagement operations"}",
  "name": "usermanagement",
  "state": "{"type":"postgresql","schema":"usermanagement_schema"}",
  "things": "[{"id":"","module_id":"","name":"Allows","description":"Allows entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"new","description":"new entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"users","description":"users entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"register","description":"register entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"User","description":"User entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"must","description":"must entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"provide","description":"provide entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"email","description":"email entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"password","description":"password entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"Password","description":"Password entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"hashed","description":"hashed entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890}]",
  "type": "core",
}
`;

exports[`ModuleGenerator should generate a ui module from a domain 1`] = `
{
  "behaviors": "[{"id":"","module_id":"","name":"viewanalytic","description":"Allows users to view their fitness analytics on a dashboard","trigger":"rpc_call","input_schema":"{}","output_schema":"{}","actions":"[\\"Display charts\\",\\"Show progress graphs\\"]","created_at":1234567890,"updated_at":1234567890}]",
  "components": "[]",
  "description": "Displays user dashboard and analytics",
  "domain_id": "domain-2",
  "flows": "[]",
  "interface": "{"type":"web_app","description":"Web interface for DashboardDisplay"}",
  "name": "dashboarddisplay",
  "screens": "[]",
  "state": "{"type":"nanostores","stores":[]}",
  "things": "[{"id":"","module_id":"","name":"Allows","description":"Allows entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"users","description":"users entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"view","description":"view entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"their","description":"their entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"fitness","description":"fitness entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"analytics","description":"analytics entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"dashboard","description":"dashboard entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"Display","description":"Display entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"charts","description":"charts entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"Show","description":"Show entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"progress","description":"progress entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890},{"id":"","module_id":"","name":"graphs","description":"graphs entity","schema":"{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"format\\":\\"uuid\\"},\\"created_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"updated_at\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"}},\\"required\\":[\\"id\\"]}","invariants":"[]","relationships":"[]","created_at":1234567890,"updated_at":1234567890}]",
  "type": "ui",
}
`;
